File: connectors/cpp/tests/integration/base_integration_test.cpp
#include "base_integration_test.hpp"
#include <thread>
#include <chrono>

namespace playhouse::test {

std::unique_ptr<TestServerFixture> BaseIntegrationTest::test_server_;

void BaseIntegrationTest::SetUp() {
    // Initialize test server fixture (singleton pattern)
    if (!test_server_) {
        test_server_ = std::make_unique<TestServerFixture>();
    }

    // Create connector with default config
    connector_ = std::make_unique<Connector>();

    config_.send_buffer_size = 65536;
    config_.receive_buffer_size = 262144;
    config_.heartbeat_interval_ms = 10000;
    config_.request_timeout_ms = 5000;  // Shorter timeout for tests
    config_.enable_reconnect = false;

    connector_->Init(config_);
}

void BaseIntegrationTest::TearDown() {
    if (connector_) {
        if (connector_->IsConnected()) {
            connector_->Disconnect();
            // Wait a bit for clean disconnect
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
        connector_.reset();
    }
}

bool BaseIntegrationTest::CreateStageAndConnect(const std::string& stage_type) {
    // Create stage via HTTP API
    try {
        if (stage_type == "TestStage") {
            stage_info_ = test_server_->GetOrCreateTestStage();
        } else {
            stage_info_ = test_server_->CreateStage(stage_type);
        }
    } catch (const std::exception& e) {
        return false;
    }

    if (!stage_info_.success) {
        return false;
    }

    return ConnectAndWait(5000);
}

bool BaseIntegrationTest::WaitForConditionWithMainThreadAction(
    std::function<bool()> condition,
    int timeout_ms
) {
    auto deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(timeout_ms);

    while (!condition()) {
        if (std::chrono::steady_clock::now() >= deadline) {
            return false;
        }

        if (connector_) {
            connector_->MainThreadAction();
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    return true;
}

bool BaseIntegrationTest::ConnectAndWait(int timeout_ms) {
    if (!connector_) {
        return false;
    }

    bool connected = false;
    bool error = false;

    connector_->OnConnect = [&connected]() { connected = true; };
    connector_->OnError = [&error](int, std::string) { error = true; };

    auto future = connector_->ConnectAsync(
        test_server_->GetHost(),
        test_server_->GetTcpPort()
    );
    (void)future;

    bool finished = WaitForConditionWithMainThreadAction([&]() {
        return connected || error;
    }, timeout_ms);

    return finished && connected;
}

bool BaseIntegrationTest::RequestAndWait(Packet packet, Packet& out_response, int timeout_ms) {
    if (!connector_) {
        return false;
    }

    bool done = false;
    connector_->Request(std::move(packet), [&](Packet response) {
        out_response = std::move(response);
        done = true;
    });

    return WaitForConditionWithMainThreadAction([&]() { return done; }, timeout_ms);
}

bool BaseIntegrationTest::AuthenticateAndWait(Packet packet, bool& out_success, int timeout_ms) {
    if (!connector_) {
        return false;
    }

    bool done = false;
    connector_->Authenticate(std::move(packet), [&](bool success) {
        out_success = success;
        done = true;
    });

    return WaitForConditionWithMainThreadAction([&]() { return done; }, timeout_ms);
}

bool BaseIntegrationTest::AuthenticateTestUser(const std::string& user_id,
                                               const std::string& token,
                                               int timeout_ms) {
    Bytes payload = proto::EncodeAuthenticateRequest(user_id, token);
    Packet auth_packet = Packet::FromBytes("AuthenticateRequest", std::move(payload));
    bool auth_success = false;
    bool completed = AuthenticateAndWait(std::move(auth_packet), auth_success, timeout_ms);
    return completed && auth_success;
}

bool BaseIntegrationTest::CreateStageConnectAndAuthenticate(const std::string& user_id,
                                                            const std::string& token,
                                                            int timeout_ms) {
    if (!CreateStageAndConnect()) {
        return false;
    }
    return AuthenticateTestUser(user_id, token, timeout_ms);
}

TestServerFixture& BaseIntegrationTest::GetTestServer() {
    if (!test_server_) {
        test_server_ = std::make_unique<TestServerFixture>();
    }
    return *test_server_;
}

} // namespace playhouse::test

File: connectors/javascript/tests/integration/helpers/BaseIntegrationTest.ts
/**
 * Base Integration Test Class
 *
 * Provides common setup, teardown, and helper methods for integration tests.
 */

import { beforeEach, afterEach } from 'vitest';
import { Connector } from '../../../src/connector.js';
import { Packet } from '../../../src/packet.js';
import { TestServerClient, CreateStageResponse } from './TestServerClient.js';
import {
    serializeAuthenticateRequest,
    parseAuthenticateReply,
    serializeEchoRequest,
    parseEchoReply,
    parseFailReply,
    parseBroadcastNotify,
    AuthenticateReply,
    EchoReply,
    FailReply,
    BroadcastNotify
} from './TestMessages.js';

/**
 * Protobuf message interface
 */
export interface IProtobufMessage {
    [key: string]: any;
}

/**
 * Base class for integration tests
 *
 * Provides:
 * - Connector lifecycle management
 * - Test server client
 * - Helper methods for common operations
 */
export class BaseIntegrationTest {
    protected testServer: TestServerClient;
    protected connector: Connector | null = null;
    protected stageInfo: CreateStageResponse | null = null;

    constructor() {
        this.testServer = new TestServerClient();
    }

    /**
     * Setup before each test
     */
    protected async beforeEach(): Promise<void> {
        this.connector = new Connector();
        this.connector.init({
            requestTimeoutMs: 5000,
            heartbeatIntervalMs: 10000
        });
    }

    /**
     * Cleanup after each test
     */
    protected async afterEach(): Promise<void> {
        if (this.connector) {
            if (this.connector.isConnected) {
                this.connector.disconnect();
                await this.delay(100);
            }
            this.connector = null;
        }
        this.stageInfo = null;
    }

    /**
     * Create a stage and connect to it
     * @param stageType Stage type to create
     * @returns Connection success
     */
    protected async createStageAndConnect(stageType: string = 'TestStage'): Promise<boolean> {
        this.stageInfo = await this.testServer.createStage(stageType);

        if (!this.connector) {
            throw new Error('Connector not initialized');
        }

        const wsUrl = `${this.testServer.wsUrl}`;
        await this.connector.connect(wsUrl, BigInt(this.stageInfo.stageId));
        return this.connector.isConnected;
    }

    /**
     * Authenticate with the test server (using protobuf format)
     * @param userId User ID for authentication
     * @param token Authentication token
     * @returns Authentication reply
     */
    protected async authenticate(userId: string, token: string = 'valid_token'): Promise<AuthenticateReply> {
        if (!this.connector) {
            throw new Error('Connector not initialized');
        }

        const payload = serializeAuthenticateRequest({ userId, token });
        const requestPacket = Packet.fromBytes('AuthenticateRequest', payload);
        const responsePacket = await this.connector.authenticate(requestPacket);

        return parseAuthenticateReply(responsePacket.payload);
    }

    /**
     * Send an echo request (using protobuf format)
     * @param content Content to echo
     * @param sequence Sequence number
     * @returns Echo reply
     */
    protected async echo(content: string, sequence: number = 1): Promise<EchoReply> {
        if (!this.connector) {
            throw new Error('Connector not initialized');
        }

        const payload = serializeEchoRequest({ content, sequence });
        const requestPacket = Packet.fromBytes('EchoRequest', payload);
        const responsePacket = await this.connector.request(requestPacket);

        return parseEchoReply(responsePacket.payload);
    }

    /**
     * Parse packet payload as a protobuf AuthenticateReply
     * @param packet Packet to parse
     * @returns Parsed AuthenticateReply
     */
    protected parseAuthReply(packet: Packet): AuthenticateReply {
        return parseAuthenticateReply(packet.payload);
    }

    /**
     * Parse packet payload as a protobuf EchoReply
     * @param packet Packet to parse
     * @returns Parsed EchoReply
     */
    protected parseEchoReply(packet: Packet): EchoReply {
        return parseEchoReply(packet.payload);
    }

    /**
     * Parse packet payload as a generic object (using protobuf parsers)
     * @param packet Packet to parse
     * @returns Parsed payload
     */
    protected parsePayload(packet: Packet): IProtobufMessage {
        try {
            switch (packet.msgId) {
                case 'AuthenticateReply':
                    return parseAuthenticateReply(packet.payload) as unknown as IProtobufMessage;
                case 'EchoReply':
                    return parseEchoReply(packet.payload) as unknown as IProtobufMessage;
                case 'FailReply':
                    return parseFailReply(packet.payload) as unknown as IProtobufMessage;
                case 'BroadcastNotify':
                    return parseBroadcastNotify(packet.payload) as unknown as IProtobufMessage;
                default:
                    // Unknown message type, try JSON fallback
                    if (packet.payload && packet.payload.length > 0) {
                        return JSON.parse(new TextDecoder().decode(packet.payload));
                    }
            }
        } catch (e) {
            console.error(`Failed to parse ${packet.msgId}:`, e);
        }
        return {};
    }

    /**
     * Wait for a condition to become true
     * @param condition Function that returns true when done
     * @param timeoutMs Maximum time to wait in milliseconds
     * @returns True if condition met, false if timeout
     */
    protected async waitForCondition(
        condition: () => boolean,
        timeoutMs: number = 5000
    ): Promise<boolean> {
        const startTime = Date.now();

        while (!condition()) {
            if (Date.now() - startTime > timeoutMs) {
                return false;
            }
            await this.delay(10);
        }

        return true;
    }

    /**
     * Wait for a condition with mainThreadAction calls
     * @param condition Function that returns true when done
     * @param timeoutMs Maximum time to wait in milliseconds
     * @returns True if condition met, false if timeout
     */
    protected async waitForConditionWithMainThread(
        condition: () => boolean,
        timeoutMs: number = 5000
    ): Promise<boolean> {
        const startTime = Date.now();

        while (!condition()) {
            if (Date.now() - startTime > timeoutMs) {
                return false;
            }
            this.connector?.mainThreadAction();
            await this.delay(10);
        }

        return true;
    }

    /**
     * Wait for a promise with mainThreadAction calls
     * @param promise Promise to wait for
     * @param timeoutMs Maximum time to wait in milliseconds
     * @returns Promise result
     */
    protected async waitWithMainThreadAction<T>(
        promise: Promise<T>,
        timeoutMs: number = 5000
    ): Promise<T> {
        const startTime = Date.now();
        let resolved = false;
        let result: T;
        let error: any;

        promise.then(
            (value) => {
                resolved = true;
                result = value;
            },
            (err) => {
                resolved = true;
                error = err;
            }
        );

        while (!resolved) {
            if (Date.now() - startTime > timeoutMs) {
                throw new Error(`Operation timed out after ${timeoutMs}ms`);
            }
            this.connector?.mainThreadAction();
            await this.delay(10);
        }

        if (error) {
            throw error;
        }

        return result!;
    }

    /**
     * Delay helper
     * @param ms Milliseconds to delay
     */
    protected async delay(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}
